src/noninteractive.rs

```diff
@@ -0,0 +1,152 @@
+//! Module containing functionality for serializing tokens and using them at a later point for
+//! non-interactive services.
+use crate::authenticator::Authenticator;
+use crate::error::Error;
+use crate::refresh::RefreshFlow;
+use crate::types::{ApplicationSecret, TokenInfo};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Clone, Default, Debug)]
+struct Entry {
+    scopes: Vec<String>,
+    refresh_token: String,
+}
+
+impl Entry {
+    fn create<T>(scopes: &[T], refresh_token: String) -> Self
+    where
+        T: AsRef<str>,
+    {
+        Entry {
+            scopes: (scopes.iter().map(|x| x.as_ref().to_string()).collect()),
+            refresh_token,
+        }
+    }
+
+    fn is_subset<T>(&self, scopes: &[T]) -> bool
+    where
+        T: AsRef<str>,
+    {
+        scopes
+            .iter()
+            .all(|scope| self.scopes.iter().any(|s| s.as_str() == scope.as_ref()))
+    }
+}
+
+/// These tokens are meant to be constructed interactively using another flow, and then can be
+/// serialized to be deserialized and used non-interactively later on.  Since access tokens are
+/// typically short-lived, this authenticator assumes it will be expired and only stores the
+/// refresh token.
+#[derive(Deserialize, Serialize, Clone, Default, Debug)]
+pub struct NoninteractiveTokens {
```

Comment:
It may be that you have good reasons to implement this, but it looks similar to `storage::JSONTokens`. Maybe you can reuse that type, as the storage requirements seem to be very similar (store/retrieve by scope).